common/map_store.c

```diff
@@ -0,0 +1,504 @@
+	/* Nodes have to be in order */
+	if (node_id_cmp(mapped_node_get_id(map, n[0]),
+			mapped_node_get_id(map, n[1])) < 0)
+		new_channel(map, &scid, n[0], n[1]);
+	else
+		new_channel(map, &scid, n[1], n[0]);
+}
+
+/* BOLT #7:
+ * 1. type: 258 (`channel_update`)
+ * 2. data:
+ *     * [`signature`:`signature`]
+ *     * [`chain_hash`:`chain_hash`]
+ *     * [`short_channel_id`:`short_channel_id`]
+ *     * [`u32`:`timestamp`]
+ *     * [`byte`:`message_flags`]
+ *     * [`byte`:`channel_flags`]
+ *     * [`u16`:`cltv_expiry_delta`]
+ *     * [`u64`:`htlc_minimum_msat`]
+ *     * [`u32`:`fee_base_msat`]
+ *     * [`u32`:`fee_proportional_millionths`]
+ *     * [`u64`:`htlc_maximum_msat`] (option_channel_htlc_max)
+ */
+static void update_channel(struct gossip_map *map, const u8 *c)
+{
+	const struct cupdate {
+		u8 sig[64];
+		u8 chainhash[32];
+		be64 scid;
+		be32 timetamp;
+		u8 message_flags, channel_flags;
+		be16 cltv_expiry_delta;
+		be64 htlc_minimum_msat;
+		be32 fee_base_msat;
+		be32 fee_proportional_millionths;
+		be64 htlc_maximum_msat;
+	} *cupdate = (void *)(c + 2); /* Skip type */
+	struct short_channel_id scid;
+	struct mapped_chan *chan;
+	struct half_chan *hc;
+
+	scid.u64 = be64_to_cpu(cupdate->scid);
+	chan = find_chan_in_map(map, &scid);
+	if (!chan)
+		errx(1, "update for channel %s not found!",
+		     type_to_string(tmpctx, struct short_channel_id, &scid));
+
+	hc = &chan->half[cupdate->channel_flags & 1];
+
+	/* Disabled flag? */
+	if (cupdate->channel_flags & 2)
+		goto disable;
+
+	hc->htlc_min = u64_to_fixedp(be64_to_cpu(cupdate->htlc_minimum_msat),
+				     true);
+	/* I checked my node: 60189 of 62358 channel_update have
+	 * htlc_maximum_msat, so we don't bother setting the rest to the
+	 * channel size (which we don't even read from the gossip_store, let
+	 * alone give up precious bytes to remember) */
+	if (cupdate->message_flags & 1)
+		hc->htlc_max
+			= u64_to_fixedp(be64_to_cpu(cupdate->htlc_maximum_msat),
+					true);
+	else
+		hc->htlc_max = 0xFFFF;
+	hc->base_fee = be32_to_cpu(cupdate->fee_base_msat);
+	hc->proportional_fee = be32_to_cpu(cupdate->fee_proportional_millionths);
+	hc->delay = be16_to_cpu(cupdate->cltv_expiry_delta);
+
+	/* Check they fit */
+	if (hc->base_fee != be32_to_cpu(cupdate->fee_base_msat)
+	    || hc->proportional_fee != be32_to_cpu(cupdate->fee_proportional_millionths)
+	    || hc->delay != be16_to_cpu(cupdate->cltv_expiry_delta)) {
+		warnx("channel_update %s ignored: fee %u/%u cltv %u too large",
+		      type_to_string(tmpctx, struct short_channel_id, &scid),
+		      be32_to_cpu(cupdate->fee_base_msat),
+		      be32_to_cpu(cupdate->fee_proportional_millionths),
+		      be16_to_cpu(cupdate->cltv_expiry_delta));
+		goto disable;
+	}
+	return;
+
+disable:
+	hc->htlc_min
+		= hc->htlc_max
+		= hc->base_fee
+		= hc->proportional_fee
+		= hc->delay
+		= 0;
+}
+
+static void map_catchup(struct gossip_map *map)
+{
+	size_t reclen;
+
+	/* You can handle unaligned accesses, right? */
```

Comment:
The "correct" fix for unaligned access is to have a local variable that you `memcpy` to:

    struct gossip_hdr ghdr_stack;
    memcpy(&ghdr_stack, (void *)(map->map + map->map_end), sizeof(struct gossip_hdr));
    struct gossip_hdr *ghdr = &ghdr_stack;

Of course, determining if unaligned access is allowed is hard.  My usual habit is to always use the above `memcpy` trick.

Reply:
We already have an HAVE_UNALIGNED_ACCESS macro. But there's also the case where we can't mmap and need to read(). The trick I've used in the past is to have a function which returns a direct pointer if it can, or does a copy and returns a pointer to that. Just need to plumb it in...